File: solutions/code-carl/ch08_greedy/T6_LC0045_Jump.java
/**
 * Practice template for:
 * - LeetCode 45. Jump Game II
 *   https://leetcode.cn/problems/jump-game-ii/
 */
public class T6_LC0045_Jump {
    public int jump(int[] nums) {
        if (nums.length == 1) return 0;
        int currCover = 0;
        int nextCover = 0;
        int count = 0;

        for (int i = 0; i <= currCover; i++) {
            nextCover = Math.max(nextCover, i + nums[i]);
            if (nextCover >= nums.length - 1) {
                count++;
                return count;
            }
            if ( i == currCover) {
                count++;
                currCover = nextCover;
            }
        }
        return -1;
    }
}

File: solutions/python/code-carl/ch08_greedy/t6_lc0045_jump.py
"""LeetCode 45. Jump Game II
https://leetcode.cn/problems/jump-game-ii/
"""

from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        if len(nums) <= 1:
            return 0

        curr_cover: int = 0
        next_cover: int = 0
        cnt: int = 0
        i: int = 0

        while i <= curr_cover:
            next_cover = max(next_cover, i + nums[i])
            if next_cover >= len(nums) - 1:
                return cnt + 1
            elif i == curr_cover:
                curr_cover = next_cover
                cnt += 1
            i += 1

        return cnt


if __name__ == "__main__":
    solution = Solution()
    assert solution.jump([2, 3, 1, 1, 4]) == 2
    assert solution.jump([2, 3, 0, 1, 4]) == 2
    print("All tests passed.")
